enhancements/csvless-bundles.md

```diff
@@ -24,209 +24,488 @@ status: implementable
+```
+
+Add a new command to `opm`, `opm bundle build`, which takes as input a directory tree -- described in [File Format](#file-format) -- and outputs an Operator Bundle image.
+
+_Ex._
 
-#### Required Manifests
+```sh
+$ opm bundle build . # detect the mediatype of the current directory, build it, and if successful, write the image tar.gz to stdout
+$ opm bundle build -o image://quay.io/my/bundle:v1.0.0 . # detect the mediatype of the current directory, build it, and if successful, push the image to a remote reference
+```
+#### Additively Build an Index
 
-The absence of a CSV means that there is a set of manifests that must ALWAYS be present to build `registry+v2` bundles. These manifests are ALWAYS __required__:
+As a Pipeline Maintainer, I want to additively build an index of `registry+v2` bundles so that I can release new bundles to an existing index.
 
-- a `Deployment` for the operator
-- a `metadata/olm.yaml` file with the following fields
-  - name
-  - version
-  - minKubeVersion
-  - installModes
+##### UX
 
-In certain situations, more manifests are required:
+Extend the `opm index add` command to accept `registry+v2` bundles.
 
-- providing an API from an APIService requires
-  - an `APIService`
-  - the `Service` it references
-- providing an admission webhook
-  - a `MutatingWebhookConfiguration` or a `ValidatingAdmissionWebhook`
-  - the `Service` it references
+#### Direct Application
 
-Other manifests are `optional`:
+As an Operator Author, I want a way to directly apply a `registry+v2` bundle to a cluster -- without a `Subscription` -- so that I can:
 
-- `Roles`, `RoleBindings`, `ClusterRoles`, and `ClusterRolesBindings`
-- `CustomResourceDefinitions`
+- validate the bundle's content
+- rapidly prototype my operator
 
-#### `opm` Support
+##### UX
 
-[`opm`](https://github.com/operator-framework/operator-registry#overview) must be updated to provide support for building `registry+v2` bundles. This can be acheived with explicit flags or through CSV/olm.yaml file presence.
+Add a new command to `opm`, `opm bundle crds`, which takes as input an Operator Bundle and outputs the set of kubectl-able CustomResourceDefinitions (CRDs) it provides.
 
-#### CSV Synthesis
+```sh
+$ opm bundle crds . # detect the mediatype of the current directory and write a YAML stream containing the provided CRDs to stdout
+$ opm bundle crds -o crds.yaml image://quay.io/my/bundle:v1.0.0 # pull and unpack the bundle image, then write the YAML stream to a file
+```
 
-OLM will support `registry+v2` bundles by synthesizing a CSV from its content. Resources usually generated by a CSV (e.g. APIService and Service) will not be directly applied, but instead be projected onto a synthesized CSV.
+Add a new command to `opm`, `opm bundle resources`, which takes as input an Operator Bundle and outputs a set of kubectl-able manifests that represent what OLM would apply to the cluster -- sans provided CRDs and dependencies-- for that bundle.
```

Comment:
Not sure I love the word `resources` for this but can't suggest a better term at the moment.

Also, I know we don't use verbs in `opm` but this seems like a good use case for `get` ie `opm get bundle resources`

Reply:
We could also combine the two and leave it up to the user to split the stream into crds and other (maybe using jq/faq).

Reply:
I like that idea. Seems like something like `opm list bundle objects` in that case

Reply:
Shouldn't it be `opm bundle list <bundle_path>`.

Then you could have something like `opm bundle list image://quay.io/my/bundle:v1.0.0 -o json | jq 'map(select(.apiVersion == "apiextensions.k8s.io/v1" and .kind == "CustomResourceDefinition")) | kubectl apply -`

Reply:
That looks good to me. Although I think eventually we probably want to provide more knobs on top of `opm bundle list` to make it easier for users to get the manifests they actually want to apply out. I know using multiple small tools like `opm` and `jq` is nice and follows the unix philosophy but for users not familiar with these tools it can be frustrating. Initially though this works fine.